File: app/src/biota/decode.ts
// Biota decoder: PNG pixels -> text. Port of decoder/decode.py.
//   composite_lum -> estimate scale (axial autocorrelation) -> matched-filter
//   tile centers -> snap to lattice -> classify rotations (NCC vs 6 templates)
//   -> base-6 digits -> bytes -> Reed-Solomon -> decrypt + verify MAC.

import { fft, fft2d, nextPow2 } from "./fft";
import {
  RGBA, Gray, compositeLum, resizeGray, centerCrop, diskMask, maxFilter2D,
} from "./imaging";
import { digitsToBytes } from "./digits";
import { rsDecodeFrame } from "./rs";
import { decryptPayload, bytesToString } from "./crypto";
import { DX0, DY0, R0 } from "./constants";

const MAXWORK = 2000; // downscale loaded images so the FFTs stay bounded

export interface TemplateLum { gray: Gray } // native-resolution luminance of tpl_k
export type TemplateSet = TemplateLum[]; // length 6

function median(arr: number[]): number {
  if (arr.length === 0) return 0;
  const a = arr.slice().sort((x, y) => x - y);
  const m = a.length >> 1;
  return a.length % 2 ? a[m] : (a[m - 1] + a[m]) / 2;
}

function mean(a: Float32Array): number {
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i];
  return s / a.length;
}

// ---------- scale estimation via axial autocorrelation ----------
function axisFund(sig: Float64Array, lim: number): { f: number | null; conf: number } {
  const peaks: number[] = [];
  for (let r = 15; r < lim - 1; r++) if (sig[r] >= sig[r - 1] && sig[r] > sig[r + 1] && sig[r] > 0) peaks.push(r);
  if (peaks.length === 0) return { f: null, conf: 0 };
  let mx = -Infinity;
  for (const r of peaks) if (sig[r] > mx) mx = sig[r];
  const strong = peaks.filter((r) => sig[r] >= 0.6 * mx);
  const f = Math.min(...strong);
  return { f, conf: sig[f] / sig[0] };
}

// accumulate Σ |FFT(line - mean)|^2 over rows (axis=0) or columns (axis=1), IFFT -> autocorr
function axialAutocorr(F: Float32Array, w: number, h: number, mu: number, axis: 0 | 1): Float64Array {
  const len = axis === 0 ? w : h;
  const cnt = axis === 0 ? h : w;
  const P = nextPow2(2 * len);
  const acc = new Float64Array(P);
  const re = new Float64Array(P), im = new Float64Array(P);
  for (let c = 0; c < cnt; c++) {
    re.fill(0); im.fill(0);
    if (axis === 0) {
      const off = c * w;
      for (let x = 0; x < len; x++) re[x] = F[off + x] - mu;
    } else {
      for (let y = 0; y < len; y++) re[y] = F[y * w + c] - mu;
    }
    fft(re, im, false);
    for (let k = 0; k < P; k++) acc[k] += re[k] * re[k] + im[k] * im[k];
  }
  const ai = new Float64Array(P);
  fft(acc, ai, true); // inverse; acc now holds the real autocorrelation
  return acc;
}

function estimateScale(F: Float32Array, w: number, h: number): number {
  const mu = mean(F);
  const ah = axialAutocorr(F, w, h, mu, 0);
  const av = axialAutocorr(F, w, h, mu, 1);
  const rh = axisFund(ah, Math.min(1500, w >> 1));
  const rv = axisFund(av, Math.min(1500, h >> 1));
  const cand: [number, number][] = []; // [conf, scale]
  if (rh.f) cand.push([rh.conf, rh.f / DX0]);
  if (rv.f) cand.push([rv.conf, rv.f / (2 * DY0)]);
  if (cand.length === 0) return 1;
  cand.sort((a, b) => b[0] - a[0]);
  return cand[0][1];
}

// ---------- matched filter (FFT convolution, 'same') ----------
// Returns cross-correlation of F0 (w x h) with an (s x s) odd kernel, centered.
function matchedFilter(F0: Float32Array, w: number, h: number, kern: Float32Array, s: number): Float32Array {
  const r = s >> 1;
  const Pw = nextPow2(w + s - 1), Ph = nextPow2(h + s - 1);
  const re = new Float64Array(Pw * Ph), im = new Float64Array(Pw * Ph);
  for (let y = 0; y < h; y++) for (let x = 0; x < w; x++) re[y * Pw + x] = F0[y * w + x];
  // kernel reversed in both axes (conv with reversed kernel == correlation)
  const kre = new Float64Array(Pw * Ph), kim = new Float64Array(Pw * Ph);
  for (let y = 0; y < s; y++) for (let x = 0; x < s; x++) kre[y * Pw + x] = kern[(s - 1 - y) * s + (s - 1 - x)];
  fft2d(re, im, Pw, Ph, false);
  fft2d(kre, kim, Pw, Ph, false);
  for (let i = 0; i < re.length; i++) {
    const a = re[i], b = im[i], c = kre[i], d = kim[i];
    re[i] = a * c - b * d;
    im[i] = a * d + b * c;
  }
  fft2d(re, im, Pw, Ph, true);
  const out = new Float32Array(w * h);
  for (let p = 0; p < h; p++) for (let q = 0; q < w; q++) out[p * w + q] = re[(p + r) * Pw + (q + r)];
  return out;
}

// ---------- classification ----------
interface Classifier {
  s: number;
  mask: Uint8Array;
  maskIdx: Int32Array; // indices where mask>0
  tplVec: Float32Array[]; // 6 mean-subtracted masked vectors
  tplNorm: number[];
}

function buildClassifier(tpls: TemplateSet, scale: number): { cls: Classifier; kern: Float32Array } {
  const s = 2 * Math.round(R0 * scale) + 1;
  const mask = diskMask(s);
  const maskIdxArr: number[] = [];
  for (let i = 0; i < mask.length; i++) if (mask[i]) maskIdxArr.push(i);
  const maskIdx = Int32Array.from(maskIdxArr);

  const masked: Float32Array[] = []; // s*s masked templates (for the invariant kernel)
  for (let k = 0; k < 6; k++) {
    const ns = Math.round(tpls[k].gray.w * scale);
    const resized = Math.abs(scale - 1) > 1e-3 ? resizeGray(tpls[k].gray, ns, ns) : tpls[k].gray;
    const cropped = centerCrop(resized, s);
    for (let i = 0; i < cropped.length; i++) cropped[i] *= mask[i];
    masked.push(cropped);
  }
  // rotation-invariant kernel = mean of 6 templates, mean-centered over mask, * mask
  const Tinv = new Float32Array(s * s);
  for (let i = 0; i < Tinv.length; i++) { let v = 0; for (let k = 0; k < 6; k++) v += masked[k][i]; Tinv[i] = v / 6; }
  let tinvMean = 0; for (const i of maskIdx) tinvMean += Tinv[i]; tinvMean /= maskIdx.length;
  const kern = new Float32Array(s * s);
  for (let i = 0; i < kern.length; i++) kern[i] = (Tinv[i] - tinvMean) * mask[i];

  // per-template masked, mean-subtracted vectors + norms
  const tplVec: Float32Array[] = [];
  const tplNorm: number[] = [];
  for (let k = 0; k < 6; k++) {
    const vec = new Float32Array(maskIdx.length);
    let m = 0;
    for (let j = 0; j < maskIdx.length; j++) { vec[j] = masked[k][maskIdx[j]]; m += vec[j]; }
    m /= maskIdx.length;
    let nrm = 0;
    for (let j = 0; j < vec.length; j++) { vec[j] -= m; nrm += vec[j] * vec[j]; }
    tplVec.push(vec);
    tplNorm.push(Math.sqrt(nrm));
  }
  return { cls: { s, mask, maskIdx, tplVec, tplNorm }, kern };
}

function classify(g: Gray, cy: number, cx: number, cls: Classifier): number {
  const { s, maskIdx, tplVec, tplNorm } = cls;
  const r = s >> 1;
  if (cy - r < 0 || cx - r < 0 || cy + r >= g.h || cx + r >= g.w) return 0;
  const pv = new Float32Array(maskIdx.length);
  let m = 0;
  for (let j = 0; j < maskIdx.length; j++) {
    const idx = maskIdx[j];
    const py = (idx / s) | 0, px = idx % s;
    const v = g.px[(cy - r + py) * g.w + (cx - r + px)];
    pv[j] = v; m += v;
  }
  m /= maskIdx.length;
  let pnrm = 0;
  for (let j = 0; j < pv.length; j++) { pv[j] -= m; pnrm += pv[j] * pv[j]; }
  pnrm = Math.sqrt(pnrm);
  let best = -2, bk = 0;
  for (let k = 0; k < 6; k++) {
    let dot = 0;
    const tv = tplVec[k];
    for (let j = 0; j < pv.length; j++) dot += pv[j] * tv[j];
    const sc = dot / (pnrm * tplNorm[k] + 1e-9);
    if (sc > best) { best = sc; bk = k; }
  }
  return bk;
}

export interface ReadResult {
  digits: number[];
  scale: number;
  dx: number;
  dy: number;
  R: number;
  C: number;
  tiles: number;
}

/** Detect the grid in a (working-resolution) luminance image and read base-6 digits. */
function readDigits(g: Gray, tpls: TemplateSet): ReadResult {
  const w = g.w, h = g.h;
  const F = new Float32Array(w * h);
  for (let i = 0; i < F.length; i++) F[i] = g.px[i] > 128 ? 1 : 0;

  const scale = estimateScale(F, w, h);
  const { cls, kern } = buildClassifier(tpls, scale);
  const s = cls.s;

  const fMean = mean(F);
  const F0 = new Float32Array(F.length);
  for (let i = 0; i < F.length; i++) F0[i] = F[i] - fMean;
  const corr = matchedFilter(F0, w, h, kern, s);

  let cmax = -Infinity;
  for (let i = 0; i < corr.length; i++) if (corr[i] > cmax) cmax = corr[i];
  let nbh = Math.round(0.85 * DX0 * scale) | 1;
  if (nbh < 1) nbh = 1;
  const mxf = maxFilter2D(corr, w, h, nbh);

  const pts: { y: number; x: number; c: number }[] = [];
  const thr = 0.4 * cmax;
  for (let y = 0; y < h; y++) for (let x = 0; x < w; x++) {
    const v = corr[y * w + x];
    if (v === mxf[y * w + x] && v > thr) pts.push({ y, x, c: v });
  }
  if (pts.length === 0) throw new Error("No tiles detected — is this an encoded Biota PNG?");

  const dx = DX0 * scale, dy = DY0 * scale, off = dx / 2;
  const ys = pts.map((p) => p.y);
  const y0 = Math.min(...ys);
  const ay0 = y0 + median(ys.map((y) => (y - y0) - dy * Math.round((y - y0) / dy)));
  const ridx = (y: number) => Math.round((y - ay0) / dy);

  // Trim sparse boundary rows: image margins can spawn a few spurious peaks above
  // or below the grid, which would shift row indices and flip the even/odd shift
  // parity (Biota always pads to full rows, so real rows are densely populated).
  const rowCount = new Map<number, number>();
  for (const p of pts) { const r = ridx(p.y); rowCount.set(r, (rowCount.get(r) || 0) + 1); }
  const maxRow = Math.max(...rowCount.values());
  const denseRows = [...rowCount.entries()].filter(([, n]) => n >= Math.max(2, 0.4 * maxRow)).map(([r]) => r);
  const rLo = Math.min(...denseRows), rHi = Math.max(...denseRows);
  const R = rHi - rLo + 1;
  const ay = ay0 + rLo * dy; // origin of (0-based) row 0 = top dense row = encoder row 0 (even/shifted)
  const parityOff = (r: number) => (r % 2 === 0 ? off : 0); // r is 0-based

  const inBlock = pts.filter((p) => { const r = ridx(p.y); return r >= rLo && r <= rHi; });
  const vals = inBlock.map((p) => p.x - parityOff(ridx(p.y) - rLo));
  const x0 = Math.min(...vals);
  let ax = x0 + median(vals.map((v) => (v - x0) - dx * Math.round((v - x0) / dx)));

  const raw = new Map<string, { r: number; cc: number; y: number; x: number; c: number }>();
  for (const p of inBlock) {
    const r = ridx(p.y) - rLo;
    const cc = Math.round((p.x - parityOff(r) - ax) / dx);
    const key = `${r},${cc}`;
    const prev = raw.get(key);
    if (!prev || p.c > prev.c) raw.set(key, { r, cc, y: p.y, x: p.x, c: p.c });
  }
  const colc = new Map<number, number>();
  for (const v of raw.values()) colc.set(v.cc, (colc.get(v.cc) || 0) + 1);
  if ((globalThis as any).process?.env?.BIOTA_DEBUG) {
    const rowc = new Map<number, number>();
    for (const v of raw.values()) rowc.set(v.r, (rowc.get(v.r) || 0) + 1);
    console.error("[dbg] R=", R, "rLo=", rLo, "rHi=", rHi, "pts=", pts.length, "rawcells=", raw.size);
    console.error("[dbg] colc=", JSON.stringify([...colc.entries()].sort((a, b) => a[0] - b[0])));
    console.error("[dbg] rowc=", JSON.stringify([...rowc.entries()].sort((a, b) => a[0] - b[0])));
  }
  const valid = [...colc.entries()].filter(([, n]) => n >= Math.max(2, 0.5 * R)).map(([c]) => c).sort((a, b) => a - b);
  if (valid.length === 0) throw new Error("No coherent grid columns found");
  const cLo = valid[0], cHi = valid[valid.length - 1];
  const C = cHi - cLo + 1;
  ax += cLo * dx;
  const cells = new Map<string, { y: number; x: number }>();
  for (const v of raw.values()) if (v.cc >= cLo && v.cc <= cHi) cells.set(`${v.r},${v.cc - cLo}`, { y: v.y, x: v.x });

  const digits: number[] = [];
  for (let r = 0; r < R; r++) {
    for (let cc = 0; cc < C; cc++) {
      const cell = cells.get(`${r},${cc}`);
      const y = cell ? cell.y : Math.round(ay + r * dy);
      const x = cell ? cell.x : Math.round(ax + parityOff(r) + cc * dx);
      digits.push(classify(g, Math.round(y), Math.round(x), cls));
    }
  }
  return { digits, scale, dx, dy, R, C, tiles: digits.length };
}

export interface DecodeResult {
  ok: boolean; // MAC verified
  text: string; // decoded text (when ok)
  errsCorrected: number; // RS byte errors corrected
  info: ReadResult;
  stage: "decoded" | "rs-failed" | "wrong-secret";
  message?: string; // human-readable status / error
}

/** Full decode: RGBA image + 6 native templates + secret -> text. */
export function decodeImage(img: RGBA, tpls: TemplateSet, secret: string): DecodeResult {
  let g = compositeLum(img);
  const longest = Math.max(g.w, g.h);
  if (longest > MAXWORK) {
    const f = MAXWORK / longest;
    g = resizeGray(g, Math.max(1, Math.round(g.w * f)), Math.max(1, Math.round(g.h * f)));
  }
  const info = readDigits(g, tpls);
  const allb = digitsToBytes(info.digits);
  let payload: number[];
  let errs: number;
  try {
    const dec = rsDecodeFrame(allb);
    payload = dec.payload;
    errs = dec.errs;
  } catch (e) {
    return { ok: false, text: "", errsCorrected: 0, info, stage: "rs-failed", message: String((e as Error).message || e) };
  }
  const { ok, bytes } = decryptPayload(payload, secret);
  if (!ok) return { ok: false, text: "", errsCorrected: errs, info, stage: "wrong-secret", message: "Wrong secret or corrupted data (MAC mismatch)" };
  return { ok: true, text: bytesToString(bytes), errsCorrected: errs, info, stage: "decoded" };
}

File: decoder/decode.py
#!/usr/bin/env python3
"""
Decode text hidden in tile rotations of a honeycomb grid PNG.

Pipeline (reverse of the Sketch encoder):
  PNG -> detect lattice (scale/origin/size, auto) -> classify each tile's
  rotation against 6 templates -> base-6 digits -> bytes ->
  Reed-Solomon error correction -> decrypt(secret) + verify MAC -> text.

Usage:
  python3 decode.py <grid.png> "<secret>"     # direct
  python3 decode.py                            # interactive (macOS file picker + hidden secret)

Requirements: numpy scipy pillow reedsolo
Templates: tpl_0.png .. tpl_5.png (the symbol rendered at 0/60/.../300 deg, @1x)
           looked up in $HEX_TPL_DIR, ./templates, ../templates, script dir, /tmp/hexdecode.
"""
import os, sys, getpass, subprocess, hashlib
import numpy as np
from PIL import Image
from scipy.signal import fftconvolve
from scipy.ndimage import maximum_filter
try:
    import reedsolo
except Exception:
    reedsolo = None

# ---- format constants (MUST match the encoder) ----
ITER = 20000
NSYM = 8        # RS parity per body block (corrects up to 4 byte errors / block)
KBLK = 223      # RS data bytes per body block (KBLK+NSYM <= 255)
HDR_NSYM = 4    # RS parity for the 2-byte length header (corrects up to 2 errors)

# ---- intrinsic grid geometry (tile W=340 at base scale) ----
W0 = 340
DX0 = round(W0 * np.sqrt(3) / 2)   # 294
DY0 = W0 * 0.75                    # 255
R0 = 120                           # central matching-disk radius (base scale)

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
def find_tpl_dir():
    candidates = (
        os.environ.get('HEX_TPL_DIR'),
        os.path.join(SCRIPT_DIR, 'templates'),
        os.path.join(SCRIPT_DIR, '..', 'templates'),
        SCRIPT_DIR,
        '/tmp/hexdecode',
    )
    for d in candidates:
        if d and os.path.exists(os.path.join(d, 'tpl_0.png')):
            return os.path.abspath(d)
    return None
TPLDIR = find_tpl_dir()

# ========== crypto (mirrors the Sketch encoder) ==========
def sha(b): return hashlib.sha256(bytes(b)).digest()
def derive(secret, salt):
    sb = secret.encode('utf-8'); k = sha(sb + bytes(salt))
    for _ in range(ITER): k = sha(k + sb)
    return k
def keystream(key, n):
    out = b''; c = 0
    while len(out) < n:
        out += sha(key + bytes([(c >> 24) & 255, (c >> 16) & 255, (c >> 8) & 255, c & 255])); c += 1
    return out[:n]
def decrypt_payload(payload, secret):
    if len(payload) < 15: return False, b''
    N = (payload[1] << 8) | payload[2]
    salt, mac, ct = payload[3:11], payload[11:15], payload[15:15 + N]
    if len(ct) < N: return False, b''
    key = derive(secret, salt)
    ok = sha(key + ct)[:4] == mac
    pt = bytes(c ^ k for c, k in zip(ct, keystream(key, N)))
    return ok, pt

# ========== Reed-Solomon framing (decode/correct) ==========
def rs_decode_frame(allb):
    """returns (payload_bytes, errors_corrected); raises on uncorrectable/malformed input"""
    if reedsolo is None:
        raise RuntimeError("reedsolo not installed (pip install reedsolo)")
    if len(allb) < 6:
        raise ValueError("frame too short")
    h, _, he = reedsolo.RSCodec(HDR_NSYM).decode(bytearray(allb[:6]))
    bl = (h[0] << 8) | h[1]
    body = bytearray(allb[6:6 + bl])
    if len(body) < bl:
        raise ValueError(f"truncated body ({len(body)}<{bl})")
    rsc = reedsolo.RSCodec(NSYM)
    payload = bytearray(); pos = 0; errs = len(he)
    while pos < bl:
        rem = bl - pos
        bln = (KBLK + NSYM) if rem > (KBLK + NSYM) else rem
        d, _, e = rsc.decode(bytearray(body[pos:pos + bln]))
        payload += d; errs += len(e); pos += bln
    return bytes(payload), errs

# ========== image / templates ==========
def composite_lum(path):
    a = np.asarray(Image.open(path).convert('RGBA')).astype(np.float32)
    al = a[..., 3:4] / 255.0
    return (a[..., :3] * al).mean(2)          # arcs ~255, bg 0

def disk_mask(s):
    r = s // 2; yy, xx = np.ogrid[:s, :s]
    return ((yy - r) ** 2 + (xx - r) ** 2) <= r * r

def center_crop(img, s):
    h, w = img.shape; r = s // 2
    return img[h // 2 - r:h // 2 + r + 1, w // 2 - r:w // 2 + r + 1]

def load_templates(scale):
    if TPLDIR is None:
        raise FileNotFoundError("templates tpl_0..5.png not found (set $HEX_TPL_DIR)")
    s = 2 * int(round(R0 * scale)) + 1; m = disk_mask(s); tpls = []
    for k in range(6):
        t = composite_lum(os.path.join(TPLDIR, f'tpl_{k}.png'))
        if abs(scale - 1.0) > 1e-3:
            ns = int(round(t.shape[0] * scale)); t = np.asarray(Image.fromarray(t).resize((ns, ns), Image.BILINEAR))
        tpls.append(center_crop(t, s).astype(np.float32) * m)
    return tpls, m, s

def estimate_scale(F):
    F0 = F - F.mean()
    A = np.fft.fftshift(np.fft.ifft2(np.abs(np.fft.fft2(F0)) ** 2).real)
    H, W = A.shape; cy, cx = H // 2, W // 2
    def axis_fund(sig):
        pk = [r for r in range(15, len(sig) - 1) if sig[r] >= sig[r - 1] and sig[r] > sig[r + 1] and sig[r] > 0]
        if not pk: return None, 0.0
        mx = max(sig[r] for r in pk); strong = [r for r in pk if sig[r] >= 0.6 * mx]
        f = min(strong); return f, float(sig[f] / sig[0])
    fh, ch = axis_fund(A[cy, cx:cx + min(1500, W // 2)])         # horizontal -> dx
    fv, cv = axis_fund(A[cy:cy + min(1500, H // 2), cx])          # vertical -> 2*dy
    cand = []
    if fh: cand.append((ch, fh / DX0))
    if fv: cand.append((cv, fv / (2 * DY0)))
    cand.sort(reverse=True)
    return cand[0][1] if cand else 1.0

def classify(g, cy, cx, tpls, m, s):
    r = s // 2; patch = g[cy - r:cy + r + 1, cx - r:cx + r + 1]
    if patch.shape != (s, s): return 0
    pv = (patch.astype(np.float32) * m)[m > 0]; pv = pv - pv.mean()
    best, bk = -2.0, 0
    for k in range(6):
        tv = tpls[k][m > 0]; tv = tv - tv.mean()
        sc = float((pv * tv).sum() / (np.linalg.norm(pv) * np.linalg.norm(tv) + 1e-9))
        if sc > best: best, bk = sc, k
    return bk

def read_digits(grid):
    g = composite_lum(grid); F = (g > 128).astype(np.float32)
    scale = estimate_scale(F); print(f'scale est: {scale:.3f}')
    tpls, m, s = load_templates(scale)
    Tinv = np.mean(tpls, axis=0); kern = (Tinv - Tinv[m > 0].mean()) * m
    corr = fftconvolve(F - F.mean(), kern[::-1, ::-1], mode='same')
    nbh = int(round(0.85 * DX0 * scale)) | 1
    mxf = maximum_filter(corr, size=nbh)
    peaks = np.argwhere((corr == mxf) & (corr > 0.4 * corr.max()))
    pts = [(int(y), int(x), float(corr[y, x])) for y, x in peaks]
    if not pts:
        raise SystemExit('No tiles detected — is this an encoded grid PNG?')

    # geometry is fixed by the scheme: derive dx/dy/off from detected scale (no drift)
    dx = DX0 * scale; dy = DY0 * scale; off = dx / 2.0
    ys = [p[0] for p in pts]; y0 = min(ys)
    ay0 = y0 + float(np.median([(y - y0) - dy * round((y - y0) / dy) for y in ys]))
    ridx = lambda y: int(round((y - ay0) / dy))
    # trim sparse boundary rows: image margins can spawn spurious peaks that shift
    # row indices and flip the even/odd shift parity (Biota pads to full rows).
    rowc = {}
    for y, x, c in pts: rowc[ridx(y)] = rowc.get(ridx(y), 0) + 1
    maxrow = max(rowc.values())
    dense = [r for r, n in rowc.items() if n >= max(2, 0.4 * maxrow)]
    r_lo, r_hi = min(dense), max(dense)
    R = r_hi - r_lo + 1
    ay = ay0 + r_lo * dy   # origin of 0-based row 0 = top dense row = encoder row 0
    parity_off = lambda r: off if r % 2 == 0 else 0.0   # r is 0-based
    inb = [(y, x, c) for y, x, c in pts if r_lo <= ridx(y) <= r_hi]
    vals = [x - parity_off(ridx(y) - r_lo) for y, x, c in inb]; x0 = min(vals)
    ax = x0 + float(np.median([(v - x0) - dx * round((v - x0) / dx) for v in vals]))

    raw = {}
    for y, x, c in inb:
        r = ridx(y) - r_lo; cc = int(round((x - parity_off(r) - ax) / dx)); k = (r, cc)
        if k not in raw or c > raw[k][2]: raw[k] = (y, x, c)
    colc = {}
    for _, cc in raw: colc[cc] = colc.get(cc, 0) + 1
    valid = sorted(c for c, n in colc.items() if n >= max(2, 0.5 * R))
    c_lo, c_hi = valid[0], valid[-1]; C = c_hi - c_lo + 1; ax += c_lo * dx
    cells = {(r, cc - c_lo): v for (r, cc), v in raw.items() if c_lo <= cc <= c_hi}
    print(f'fit: scale={scale:.3f} dx={dx:.1f} dy={dy:.1f} R={R} C={C} cells={len(cells)}')

    digits = []
    for r in range(R):
        for cc in range(C):
            if (r, cc) in cells: y, x, _ = cells[(r, cc)]
            else: y = int(round(ay + r * dy)); x = int(round(ax + parity_off(r) + cc * dx))
            digits.append(classify(g, int(y), int(x), tpls, m, s))
    return digits

# ========== CLI ==========
def choose_file():
    try:
        out = subprocess.check_output(
            ['osascript', '-e',
             'POSIX path of (choose file with prompt "Select encoded PNG" of type {"png","public.png"})'],
            stderr=subprocess.DEVNULL)
        return out.decode().strip()
    except Exception:
        return input('Path to PNG: ').strip()

def main():
    grid = sys.argv[1] if len(sys.argv) > 1 else choose_file()
    secret = sys.argv[2] if len(sys.argv) > 2 else getpass.getpass('Secret: ')
    if not grid:
        raise SystemExit('No file selected.')

    dg = read_digits(grid)
    allb = bytes(((dg[i]*216 + dg[i+1]*36 + dg[i+2]*6 + dg[i+3]) & 0xFF) for i in range(0, len(dg) - 3, 4))
    try:
        payload, errs = rs_decode_frame(allb)
        print(f'RS: ok, corrected {errs} byte error(s)')
    except Exception as e:
        print('RS: FAILED to correct ->', e)
        print('DECODED TEXT: <unavailable: too many errors>')
        return
    ok, pt = decrypt_payload(payload, secret)
    print('secret check:', 'OK' if ok else 'FAILED (wrong secret or corrupted)')
    print('DECODED TEXT:', repr(pt.decode('utf-8', errors='replace')) if ok else '<unavailable: wrong secret>')

if __name__ == '__main__':
    main()
